x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/logical/ReplaceStatsFilteredAggWithEval.java

```diff
@@ -31,44 +34,99 @@
+                    aggHolder.set(a);
+                }
+            });
+            aggregate = aggHolder.get();
+        } else {
+            return plan; // not an Aggregate or InlineJoin, nothing to do
+        }
+
+        if (aggregate != null) {
+            int oldAggSize = aggregate.aggregates().size();
+            List<NamedExpression> newAggs = new ArrayList<>(oldAggSize);
+            List<Alias> newEvals = new ArrayList<>(oldAggSize);
+            List<NamedExpression> newProjections = new ArrayList<>(oldAggSize);
 
-        for (var ne : aggregate.aggregates()) {
-            if (ne instanceof Alias alias
-                && alias.child() instanceof AggregateFunction aggFunction
-                && aggFunction.hasFilter()
-                && aggFunction.filter() instanceof Literal literal
-                && Boolean.FALSE.equals(literal.value())) {
+            for (var ne : aggregate.aggregates()) {
+                if (ne instanceof Alias alias
+                    && alias.child() instanceof AggregateFunction aggFunction
+                    && aggFunction.hasFilter()
+                    && aggFunction.filter() instanceof Literal literal
+                    && Boolean.FALSE.equals(literal.value())) {
 
-                Object value = aggFunction instanceof Count || aggFunction instanceof CountDistinct ? 0L : null;
-                Alias newAlias = alias.replaceChild(Literal.of(aggFunction, value));
-                newEvals.add(newAlias);
-                newProjections.add(newAlias.toAttribute());
-            } else {
-                newAggs.add(ne); // agg function unchanged or grouping key
-                newProjections.add(ne.toAttribute());
+                    Object value = aggFunction instanceof Count || aggFunction instanceof CountDistinct ? 0L : null;
+                    Alias newAlias = alias.replaceChild(Literal.of(aggFunction, value));
+                    newEvals.add(newAlias);
+                    newProjections.add(newAlias.toAttribute());
+                } else {
+                    newAggs.add(ne); // agg function unchanged or grouping key
+                    newProjections.add(ne.toAttribute());
+                }
             }
-        }
 
-        LogicalPlan plan = aggregate;
-        if (newEvals.isEmpty() == false) {
-            if (newAggs.isEmpty()) { // the Aggregate node is pruned
-                plan = localRelation(aggregate.source(), newEvals);
-            } else {
-                plan = aggregate.with(aggregate.child(), aggregate.groupings(), newAggs);
-                plan = new Eval(aggregate.source(), plan, newEvals);
-                plan = new Project(aggregate.source(), plan, newProjections);
+            if (newEvals.isEmpty() == false) {
+                if (newAggs.isEmpty()) { // the Aggregate node is pruned
+                    if (ij.get() != null) { // this is an Aggregate part of right-hand side of an InlineJoin
```

Comment:
The meat of the PR. Basically, a new "branch" in dealing with pruning Aggs in InlineJoin specifically.